src/chocolatey/infrastructure.app/nuget/NugetList.cs

```diff
@@ -334,119 +321,62 @@ public static ISet<IPackageSearchMetadata> find_all_package_versions(string pack
         /// </summary>
         /// <param name="packageName">Name of package to search for</param>
         /// <param name="config">Chocolatey configuration used to help supply the search parameters</param>
-        /// <param name="nugetLogger">fill me in</param>
-        /// <param name="resources">fill me in</param>
+        /// <param name="nugetLogger">The nuget logger</param>
+        /// <param name="packageMetadataResources">The PackageMetaDataResources that should be queried</param>
+        /// <param name="listResources">The ListResources that should be queried</param>
         /// <param name="version">Version to search for</param>
         /// <param name="cacheContext">Settings for caching of results from sources</param>
         /// <returns>One result or nothing</returns>
-        public static IPackageSearchMetadata find_package(string packageName, ChocolateyConfiguration config, ILogger nugetLogger, ChocolateySourceCacheContext cacheContext, IEnumerable<PackageMetadataResource> resources, NuGetVersion version = null)
+        public static IPackageSearchMetadata find_package(
+            string packageName,
+            ChocolateyConfiguration config,
+            ILogger nugetLogger,
+            ChocolateySourceCacheContext cacheContext,
+            IEnumerable<PackageMetadataResource> packageMetadataResources,
+            IEnumerable<ListResource> listResources,
+            NuGetVersion version = null)
         {
-            if (version is null)
+            if (config.Features.UsePackageRepositoryOptimizations)
             {
-                var metadataList = new HashSet<IPackageSearchMetadata>();
-                foreach (var resource in resources)
+                if (version is null)
                 {
-                    metadataList.AddRange(resource.GetMetadataAsync(packageName, config.Prerelease, false, cacheContext, nugetLogger, CancellationToken.None).GetAwaiter().GetResult());
-                }
-                return metadataList.OrderByDescending(p => p.Identity.Version).FirstOrDefault();
-            }
+                    var metadataList = new HashSet<IPackageSearchMetadata>();
 
-            foreach (var resource in resources)
-            {
-                var metadata = resource.GetMetadataAsync(new PackageIdentity(packageName, version), cacheContext, nugetLogger, CancellationToken.None).GetAwaiter().GetResult();
-                if (metadata != null)
-                {
-                    return metadata;
+                    foreach (var listResource in listResources)
+                    {
+                        metadataList.Add(listResource.PackageAsync(packageName, config.Prerelease, nugetLogger, CancellationToken.None).GetAwaiter().GetResult());

```

Comment:
Since we don't validate whether the returned results is `null` or not, this causes a null reference exception to happen later if the package couldn't be found.

example commands
```
choco info chocolate --source chocolatey
```
(also affects outdated if one or more packages couldn't be found).

```suggestion
                        var package = listResource.PackageAsync(packageName, config.Prerelease, nugetLogger, CancellationToken.None).GetAwaiter().GetResult();
                        if (package != null)
                        {
                            metadataList.Add(package);
                        }
```

Reply:
Great catch! Will get this added!